x-pack/platform/plugins/shared/streams/server/lib/streams/system/system_client.ts

```diff
@@ -0,0 +1,184 @@
+
+interface SystemBulkIndexOperation {
+  index: { system: System };
+}
+interface SystemBulkDeleteOperation {
+  delete: { system: { name: string } };
+}
+
+function getSystemUuid(streamName: string, systemName: string) {
+  return objectHash({
+    [STREAM_NAME]: streamName,
+    [SYSTEM_NAME]: systemName,
+  });
+}
+
+function fromStorage(link: StoredSystem): System {
+  return {
+    name: link[SYSTEM_NAME],
+    description: link[SYSTEM_DESCRIPTION],
+    filter: link[SYSTEM_FILTER],
+  };
+}
+
+function toStorage(name: string, system: System): StoredSystem {
+  return {
+    [STREAM_NAME]: name,
+    [SYSTEM_UUID]: getSystemUuid(name, system.name),
+    [SYSTEM_NAME]: system.name,
+    [SYSTEM_DESCRIPTION]: system.description,
+    [SYSTEM_FILTER]: system.filter,
+  };
+}
+
+export type SystemBulkOperation = SystemBulkIndexOperation | SystemBulkDeleteOperation;
+
+export class SystemClient {
+  constructor(
+    private readonly clients: {
+      storageClient: IStorageClient<SystemStorageSettings, StoredSystem>;
+    }
+  ) {}
+
+  async syncSystemList(
+    name: string,
+    systems: System[]
+  ): Promise<{ deleted: System[]; indexed: System[] }> {
+    const systemsResponse = await this.clients.storageClient.search({
+      size: 10_000,
+      track_total_hits: false,
+      query: {
+        bool: {
+          filter: [...termQuery(STREAM_NAME, name)],
+        },
+      },
+    });
+
+    const existingSystems = systemsResponse.hits.hits.map((hit) => {
+      return hit._source;
+    });
+
+    const nextSystems = systems.map((system) => {
+      return toStorage(name, system);
+    });
+
+    const nextIds = new Set(nextSystems.map((system) => system[SYSTEM_UUID]));
+    const systemsDeleted = existingSystems.filter((system) => !nextIds.has(system[SYSTEM_UUID]));
+
+    const operations: SystemBulkOperation[] = [
+      ...systemsDeleted.map((system) => ({ delete: { system: fromStorage(system), name } })),
+      ...nextSystems.map((system) => ({ index: { system: fromStorage(system), name } })),
+    ];
+
+    if (operations.length) {
+      await this.bulk(name, operations);
+    }
+
+    return {
+      deleted: systemsDeleted.map((system) => fromStorage(system)),
+      indexed: systems,
+    };
+  }
+
+  async linkSystem(name: string, system: System): Promise<System> {
+    const document = toStorage(name, system);
+
+    await this.clients.storageClient.index({
+      id: document[SYSTEM_UUID],
+      document,
+    });
+
+    return system;
+  }
+
+  async unlinkSystem(name: string, system: System): Promise<void> {
+    const id = getSystemUuid(name, system.name);
+
+    const { result } = await this.clients.storageClient.delete({ id });
+    if (result === 'not_found') {
+      throw new SystemNotFoundError(`System ${system.name} not found for stream ${name}`);
+    }
+  }
+
+  async clean() {
+    await this.clients.storageClient.clean();
+  }
+
+  async bulk(name: string, operations: SystemBulkOperation[]) {
```

Comment:
note: This is returning lots of internals (e.g. the index name) - not sure whether we should at this point. We do the same for the assets client though and it's an internal API, so not a concern per se, but it gave me pause.

Reply:
The semantics are also a bit different, e.g. the request doesn't fail if a non-existing document is deleted.